pyacq/core/stream.py

```diff
@@ -523,11 +600,386 @@ def get_array_slice(self, index, length, autoswapaxes = True):
+        
+        # by default, index advances by the size of the chunk
+        if index is None:
+            index = self._write_index + dsize
+        
+        assert dsize <= index - self._write_index, ("Data size is %d, but index "
+                                                    "only advanced by %d." % 
+                                                    (dsize, index-self._write_index)) 
+
+        revert_inds = [self._read_index, self._write_index]
+        try:
+            # advance write index. This immediately prevents other processes from
+            # accessing memory that is about to be overwritten.
+            self._set_write_index(index)
+            
+            # decide if any skipped data needs to be filled in
+            fill_start = max(self._read_index + 1, self._write_index + 1 - bsize)
+            fill_stop = self._write_index + 1 - dsize
+            
+            if fill_stop > fill_start:
+                # data was skipped; fill in missing regions with 0.
+                self._fill(fill_start, fill_stop)
+                revert_inds[1] = fill_stop-1
+                
+            self._write(self._write_index + 1 - dsize, self._write_index + 1, data)
+                
+            self._set_read_index(index)
+        except:
+            # If there is a failure writing data, revert read/write pointers
+            self._set_read_index(revert_inds[0])
+            self._set_write_index(revert_inds[1])
+            raise
+
+    def _write(self, start, stop, value):
+        # get starting index
+        bsize = self.nbytes
+        dsize = stop - start
+        i = start % bsize
+        
+        if self.double:
+            self.buffer[i:i+dsize] = value
+            i += bsize
+        
+        if i + dsize <= self.buffer.shape[0]:
+            self.buffer[i:i+dsize] = value
+        else:
+            n = self.buffer.shape[0]-i
+            self.buffer[i:] = value[:n]
+            self.buffer[:dsize-n] = value[n:]
+            
+    def _fill(self, start, stop):
+        self._write(start, stop, 0)
+
+
+class ArrayRingBuffer(RingBuffer):
+    """Array-like object that wraps an internal ring buffer.
+    
+    This allows the user to request the concatenated history of data
+    received by the stream, up to a predefined length. Double ring buffers
+    allow faster, copyless reads at the expense of doubled write time and memory
+    footprint.    
+    """
+    def __init__(self, shape, dtype, double=True, shmem=None, fill=None):
+        self.double = double
+        self.shape = shape
+        
+        shape = (shape[0] * (2 if double else 1),) + shape[1:]
+        # initialize int buffers with 0 and float buffers with nan
+        if fill is None:
+            fill = 0 if np.dtype(dtype).kind in 'ui' else np.nan
+        self._filler = fill
+        
+        bsize = np.product(self.shape) * np.dtype(dtype).itemsize
+        RingBuffer.__init__(self, bsize, double=double, shmem=shmem)
```

Comment:
Outch, I am not sur to follow this.
I really really would like to have the 2 possibilities inetrnally (sample, channel) and (channel, sample). This is one of strong point for me in pyacq 0.2.

> Streams send / recv array chunks without altering their internal memory layout--if channels are not interleaved when they are sent, then they will not be interleaved when they are received, regardless of the stream type.

For *plaindata* this make sens but for sharredarray I would like to have all possibilities:
  1 - sender is (sample, channel) sharredarray is (sample, channel)
  2 - sender is (channel, sample) sharredarray is (sample, channel)
  3 - sender is (sample, channel) sharredarray is (channel, sample)
  4 - sender is (channel, sample) sharredarray is (channel, sample)
Case 2 and 3 imply some tranpose but numpy make it free for us because in any case externally with numpy view everything is (sample, channel). We just have swapaxis in approiate case at array creation.

> Sharedmem streams use RingBuffer to store byte-arrays in any axis order

I would prefer this approach that Sharredmem and SharredArray are agnostic to this ringbuffer mechanism.
But a RingBuffer and RingBuffer array can use sharredmem (optional in __init__), as you start to do.

> If you attach an ArrayRingBuffer to an InputStream, then by default it will use its own memory buffer and copy chunks into the buffer with timeaxis=0 as they arrive.

This is really really good, this was the missing piece in the previous design : a proxy ringbuffer for receiver even if it is not a sharredmem. But I thing we must not force internnally (sample, channel) it must be custumable. But if RingBufferArray have the possibility then we have the feature here.

> If a sharedmem stream's buffer already happens to have its memory layout with timeaxis=0, then the ArrayRingBuffer can use the mmap buffer rather than creating its own memory and copying.

I  think it must be the case event if it is (channel, sample) there are not copy when the stream is sharedmem.
If a receiver needs an internal tranpose then we put a Node in between that just do outbuffer[:] = inbuffer [:].
The internal transpose is done under the wood.